clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -422,12 +422,99 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
-static bool isZeroSizedArray(QualType Ty) {
-  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty))
-    return CATy->getSize() == 0;
+static bool isZeroSizedArray(Sema &SemaRef, QualType Ty) {
+  if (const auto *CAT = SemaRef.getASTContext().getAsConstantArrayType(Ty))
+    return CAT->getSize() == 0;
   return false;
 }
 
+// NOTE: In case this function gets upstreamed and other single source
+// offloading models also need it, it can be generalized via using of
+// `targetDiag` instead of `SYCLDiagIfDeviceCode` and removing assertion at the
+// beginning.
+void Sema::deepTypeCheckForSYCLDevice(SourceLocation UsedAt,
+                                      llvm::DenseSet<QualType> Visited,
+                                      ValueDecl *DeclToCheck) {
+  assert(getLangOpts().SYCLIsDevice &&
+         "Should only be called during SYCL compilation");
+  auto Check = [&](QualType TypeToCheck, const ValueDecl *D) {
+    bool ErrorFound = false;
+    if (isZeroSizedArray(*this, TypeToCheck)) {
+      SYCLDiagIfDeviceCode(UsedAt, diag::err_typecheck_zero_array_size);
+      ErrorFound = true;
+    }
+    // Checks for other types can also be done here.
+    if (ErrorFound) {
+      if (auto *FD = dyn_cast<FieldDecl>(D)) {
+        SYCLDiagIfDeviceCode(FD->getLocation(),
+                             diag::note_illegal_field_declared_here)
+            << FD->getType()->isPointerType() << FD->getType();
+      } else {
+        SYCLDiagIfDeviceCode(D->getLocation(),
+                             diag::note_illegal_type_decl_here)
+            << D << D->getType();
+      }
+    }
+
+    return ErrorFound;
+  };
+
+  // In case we have a Record used do the DFS for a bad field
+  SmallVector<const ValueDecl *, 4> StackForRecursion;
```

Comment:
We shouldn't be doing this level of deep checking in the CFE.  Walking the AST this deep is exactly the stuff @AaronBallman  and I have been warning against/trying to stop us from doing.

Reply:
Hmm... I guess at second look, this is going through the fields instead of calls, so it isn't as bad as I thought.

I'd still recommend doing this differently, I'd suggest diagnosing by calling RecordDecl::CheckSYCLTypes (or something, the point is that it loops through its fields and checks, plus recursing down CheckSYCLTypes when necessary), then having it return whether it diagnosed or not so that you can emit the note if you care to.

Printing EVERY error, or the history for every error is likely overly verbose anyway.

Reply:
> Hmm... I guess at second look, this is going through the fields instead of calls, so it isn't as bad as I thought.

Yes, that is going through fields. Actually I didn't invent that, it seems clang already does such thing for OpenCL kernel arguments https://github.com/intel/llvm/blob/35ccdd5a6b3c467cd39bfb14f0abc7ef2bc20667/clang/lib/Sema/SemaDecl.cpp#L8958 .

> I'd suggest diagnosing by calling RecordDecl::CheckSYCLTypes (or something, the point is that it loops through its fields and checks, plus recursing down CheckSYCLTypes when necessary), then having it return whether it diagnosed or not so that you can emit the note if you care to.

Do you mean we should stop recursion once the first field of illegal type is found?

Reply:
Ah, weird!  I'm surprised that made it by community review, though I do know that the OpenCL code owners are a bit lax as far as code review effectiveness.  That said, if instead of this patch we could just generalize the OpenCL version, at least we wouldn't be doing further harm (AND this would be a good opportunity to upstream this immediately instead of doing here).

I would probably give up on the notes after the 1st in each struct/path.  Blitting out the 'how we got here' non stop is just going to be a giant mess of notes I suspect.

Reply:
> That said, if instead of this patch we could just generalize the OpenCL version

I will try that, but I'm pretty sure this will be pretty annoying to do (if even possible!), since OpenCL code seems very... OpenCL specific.

> we wouldn't be doing further harm

If you think this does harm, I'm open for any ideas. Do the recursive search for bad struct fields is the best way to solve the problem that I could figure out.

> this would be a good opportunity to upstream this immediately instead of doing here

I believe the best way to upstream this is to ask developers of other single source offload models (CUDA/OMP/maybeHIP?) if they need such thing and generalize the function I've added across them. OpenCL folks already have their diagnostic, the generalization will likely complicate the code and I think they don't care about SYCL and therefore the change won't bring anything good for them, so this will end up rejected or it will take forever to merge.

Reply:
Yeah, the first problem with generalization is that OpenCL code emits errors for pointers met inside struct, whereas SYCL needs to take pointee type and go further instead.

Reply:
Ultimately, where we want this diagnostic to live is within SemaDecl.cpp when processing the declaration of a zero-length array. The fact that we have to wait until the entire TU is complete before deciding whether we can or cannot diagnose a declaration elsewhere is a design flaw.

Reply:
> Ultimately, where we want this diagnostic to live is within SemaDecl.cpp when processing the declaration of a zero-length array. The fact that we have to wait until the entire TU is complete before deciding whether we can or cannot diagnose a declaration elsewhere is a design flaw.

With current SYCL spec design we cannot do that. We don't know what is device code and what is not until `parallel_for` template is ready. This patch at least doesn't extend SemaSYCL kernels processing and uses deferred diagnostics like other single source offloading models.

Reply:
> With current SYCL spec design we cannot do that. We don't know what is device code and what is not until parallel_for template is ready. 

Yup, and we were asked explicitly to push back when the SYCL spec is designed in a way that's not reasonable to our implementation. CC @pvchupin for his awareness.

> This patch at least doesn't extend SemaSYCL kernels processing and uses deferred diagnostics like other single source offloading models.

I'm not blocking this review so there may not be anything actionable here for you, but the reason I'm pushing back is because this seems to be building on top of something we already know is not likely to be acceptable for upstreaming. So you're right that it doesn't make new problems, but it does make it harder for us to go back and fix that foundation. :-(

Reply:
> we already know is not likely to be acceptable for upstreaming

Why so? The whole `checkTypeSupport` routine that I'm extending is in upstream clang, and it uses the approach "emit a deferred diagnostic for unsupported type", it works not only for SYCL, but for OMP as well. The only new thing I brought here is recursive check for struct fields (using the idea that already used in community code). The reason why SYCL prohibits zero length arrays is that there is no memory allocation on GPUs (and any other weird devices where you can run opencl/sycl).

Reply:
> Why so?

As mentioned; the compiler should not have to process the entire TU before deciding whether to diagnose the declaration of some earlier construct or not. That's a deficiency of the language extension design; C++ is not a multi-pass programming language.

You piqued my interest though as to how this got into community in the first place... https://reviews.llvm.org/D109315 renamed to `checkTypeSupport()` but https://reviews.llvm.org/D74387 is what introduced the notion originally. I see John McCall was on that first review, but despite having a considerable number of comments, he never signed off on it before it landed. I didn't check all of his comments to see if this was discussed or not.

All that said, I think the *desire* for the diagnostic is reasonable and something I'm in support of. I'm pushing back on the SYCL making it impossible to implement reasonable diagnostics without having to make multiple passes over an entire TU.

Reply:
> C++ is not a multi-pass programming language.

But maybe SYCL is?

I got your point. I was just only trying to say that need to parse the entire TU to understand what is what is not only SYCL problem. Deferred diagnostics are widely used across clang, not only in `checkTypeSupport()`. I agree that SYCL could define device code entry more precisely like CUDA does with its "device/host/etc" attributes instead of "everything called from template function with this name magically becomes device code". But even CUDA has places where deferred diagnostics needed, that means even there it is not possible to understand where is device code until everything is parsed.

Reply:
> But maybe SYCL is?

SYCL is supposed to be an extension to C++ and I'd question whether redefining the compilation model is a conforming extension in practice.

> I got your point. I was just only trying to say that need to parse the entire TU to understand what is what is not only SYCL problem. Deferred diagnostics are widely used across clang, not only in checkTypeSupport(). 

I agree! The difference is: deferred diagnostics typically do not require deferring to the end of the TU, but instead to the end of the function definition or class definition (or perhaps instantiation thereof). Function bodies and class bodies are generally far more bounded than an entire TU. (Consider how well this scales to TUs that are code generated; you could get 1 million lines of code that forms an AST we have to walk multiple times only to discover that there was little or no SYCL code anywhere in the TU to begin with.)

Reply:
> SYCL is supposed to be an extension to C++

Okay, I didn't see SYCL from this point of view and rather perceived it as a separate C++ based language. 

@erichkeane , I've reduced number of emitted notes, all notes are emitted only when the first error is discovered. Please let me know if this became better.

Reply:
I'm ok with it, I'll leave the decision on the rest to the code owners.

Reply:
> > C++ is not a multi-pass programming language. [@AaronBallman]
> 
> But maybe SYCL is? [@Fznamznon]

Not really. This is an implementation detail.
 
> I got your point. I was just only trying to say that need to parse the entire TU to understand what is what is not only SYCL problem. Deferred diagnostics are widely used across clang, not only in `checkTypeSupport()`. I agree that SYCL could define device code entry more precisely like CUDA does with its "device/host/etc" attributes instead of "everything called from template function with this name magically becomes device code". 

Argh! I have a stroke! I need some oxygen! :-) I think you are ready to send your résumé to NVIDIA at that point. @brycelelbach you have a candidate! ;-)
Single-source and plain C++ without extension is the essence of SYCL.

Even handling templates in plain C++ requires some evaluation delay to provide better diagnostics.

> But even CUDA has places where deferred diagnostics needed, that means even there it is not possible to understand where is device code until everything is parsed.

And I guess that if NVIDIA is up-streaming in the future their pure NVC++ they will have a  somehow similar issue on this aspect. Any feedback from NVIDIA compiler team on this?